### PR TITLE
Laboratory work #1, Isaichkina Anastasia - 19FPL1

### PR DESCRIPTION
**Pull request by anmyfe**

### CODE REVIEW

**Comment by ValeriyaKuznetsova** [comment]
Unfortunately, the deadline is not that far. Please, push changes so that we can review them.

File: lab_1/main.py [hunk 1]
```diff
@@ -12,7 +12,17 @@ def tokenize(text: str) -> list:
     e.g. text = 'The weather is sunny, the man is happy.'
     --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
     """
-    pass
+    tokens = ""
+    text = text.lower()
+    for i in text:
+        if i not in ".,!?^/:;-%@#$^&*()_+=><[]{}\"\'\\":
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: Can you suggest a way of checking without ready-made lists?

File: lab_1/main.py [hunk 2]
```diff
@@ -25,9 +35,14 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
     stop_words = ['the', 'is']
     --> ['weather', 'sunny', 'man', 'happy']
     """
-    pass
+    for word in tokens:
+        if word in stop_words:
+            tokens.remove(word)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: try this:
stop_words = ['the', 'is']
tokens = ['the', 'is', 'a', 'the']

Does it work correctly?

File: lab_1/main.py [hunk 3]
```diff
@@ -36,9 +51,19 @@ def calculate_frequencies(tokens: list) -> dict:
     e.g. tokens = ['weather', 'sunny', 'man', 'happy']
     --> {'weather': 1, 'sunny': 1, 'man': 1, 'happy': 1}
     """
-    pass
+    freq_dict = {}
+    for key in tokens:
+        if key in freq_dict:
+            value = freq_dict[key]
+            freq_dict = freq_dict[key] + 1
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: why change the dictionary?

File: lab_1/main.py [hunk 4]
```diff
@@ -49,7 +74,25 @@ def get_top_n_words(freq_dict: dict, top_n: int) -> list:
...
+
+    sorted_dict = dict(sorted(freq_dict.items(), key=lambda item: item[1], reverse=True))
+    sorted_dict = list(sorted_dict.keys())
+
+    if top_n > len(sorted_dict):
+        top_n = len(sorted_dict)
+    if top_n >= 0:
+        while n != top_n:
+            common_words.append(sorted_dict[n])
+            n = n + 1
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: can you suggest an easier way to get top_n elements?

File: lab_1/main.py [hunk 5]
```diff
@@ -12,12 +12,17 @@ def tokenize(text: str) -> list:
...
     """
-
-    text = text.split()
+    tokens = ""
     text = text.lower()
+    for i in text:
+        if i not in ".,!?^/:;-%@#$^&*()_+=><[]{}\"\'\\":
+            tokens = tokens + i
+    tokens = tokens.split()
+    print(tokens)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: remove "print" when you push changes (everywhere)

File: lab_1/main.py [hunk 6]
```diff
@@ -35,13 +32,12 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
     stop_words = ['the', 'is']
     --> ['weather', 'sunny', 'man', 'happy']
     """
-    for word in tokens:
-        if word in stop_words:
-            tokens.remove(word)
-    return tokens
+    if type(tokens) == list:
+        if type(stop_words) == list:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: can you join lines 35-36?

File: lab_1/main.py [hunk 7]
```diff
@@ -74,27 +67,17 @@ def get_top_n_words(freq_dict: dict, top_n: int) -> list:
...
     """
-    common_words = []
-    n = 0
-
     sorted_dict = dict(sorted(freq_dict.items(), key=lambda item: item[1], reverse=True))
     sorted_dict = list(sorted_dict.keys())
 
-    if top_n > len(sorted_dict):
-        top_n = len(sorted_dict)
     if top_n >= 0:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: can you move this check to the beginning? 
if top_n is incorrect, there is no point doing anything at all

File: lab_1/main.py [hunk 8]
```diff
@@ -151,19 +131,13 @@ def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> li
     --> [['man', 'is'], ['dog, 'cat']]
     """
     concordance = []
-
-    list_index = [i for i, x in enumerate(tokens) if x == word]
-    for i in list_index:
-        if left_n > 0 and right_n > 0:
-            concordance.append([tokens[i - left_n], tokens[i + right_n]])
+    for i in get_concordance(tokens, word, left_n, right_n):
+        print(i)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: do not forget to remove

File: lab_1/start.py [hunk 9]
```diff
@@ -2,7 +2,8 @@
 Concordance implementation starter
 """
 
-from lab_1.main import read_from_file
+from main import read_from_file
```
Comment (ValeriyaKuznetsova) [inline lab_1/start.py]: do not push such changes, change it back

File: lab_1/start.py [hunk 10]
```diff
@@ -4,16 +4,42 @@
...
+    concordance = main.get_concordance(tokens, "book", 2, 3)
+    print("concordance for 'book':", concordance[:5])
+
+    adjacent_words = main.get_adjacent_words(tokens, "book", 2, 3)
+    print("adjacent words for 'book':", adjacent_words[:5])
+
+    sorted_concordance = main.sort_concordance(tokens, "book", 2, 3, False)
+    print("sorted concordance for 'book':", sorted_concordance[:3])
+
+    main.write_to_file('report.txt', sorted_concordance)
```
Comment (anmyfe) [inline lab_1/start.py]: здесь появляется ошибка "PermissionError: [Errno 13] Permission denied: 'report.txt'", не понимаю, как исправить

File: lab_1/get_adjacent_words_test.py [hunk 11]
```diff
@@ -87,7 +88,7 @@ def test_big_text_get_adjacent_words_term(self):
         """
         Checks if adjacent words for a given term can be found properly
         """
-        text = read_from_file('lab_1/data.txt')
+        text = read_from_file('data.txt')
```
Comment (ValeriyaKuznetsova) [inline lab_1/get_adjacent_words_test.py]: do not push such changes

File: lab_1/main.py [hunk 12]
```diff
@@ -12,14 +12,13 @@ def tokenize(text: str) -> list:
...
-    for i in text:
-        if i not in ".,!?^/:;-%@#$^&*()_+=><[]{}\"\'\\":
-            tokens = tokens + i
-    tokens = tokens.split()
 
-    return tokens
+    if text and isinstance(text, str):
+        tokens = re.sub('[^a-zA-Z \n]', '', text).lower()
+        return tokens.split()
+    else:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: you don't need this "else", if "return tokens.split()" works, the function will finish its work

File: lab_1/main.py [hunk 13]
```diff
@@ -32,12 +31,17 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
...
-    if type(tokens) == list:
-        if type(stop_words) == list:
-            tokens = [i for i in tokens if i not in stop_words]
+    if tokens and isinstance(tokens, list):
+        for word in tokens:
+            if not isinstance(word, str):
+                return []
 
-    return tokens
+        if stop_words and isinstance(stop_words, list):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: why here? can you add this if-statement to line 34?

File: lab_1/main.py [hunk 14]
```diff
@@ -32,12 +31,17 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
...
+            tokens = [i for i in tokens if i not in stop_words and i.isalpha]
 
+        return tokens
+    else:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: same about else

File: lab_1/main.py [hunk 15]
```diff
@@ -32,12 +31,17 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
...
-        if type(stop_words) == list:
-            tokens = [i for i in tokens if i not in stop_words]
+    if tokens and isinstance(tokens, list):
+        for word in tokens:
+            if not isinstance(word, str):
+                return []
 
-    return tokens
+        if stop_words and isinstance(stop_words, list):
+            tokens = [i for i in tokens if i not in stop_words and i.isalpha]
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: does the if statement really work? (look at i.isalpha)

File: lab_1/main.py [hunk 16]
```diff
@@ -67,13 +73,13 @@ def get_top_n_words(freq_dict: dict, top_n: int) -> list:
     top_n = 1
     --> ['happy']
     """
-    sorted_dict = dict(sorted(freq_dict.items(), key=lambda item: item[1], reverse=True))
-    sorted_dict = list(sorted_dict.keys())
 
-    if top_n >= 0:
+    common_words = []
+
+    if isinstance(freq_dict, dict) and isinstance(top_n, int) and freq_dict and top_n >= 0:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: do we need top_n = 0?

File: lab_1/main.py [hunk 17]
```diff
@@ -96,24 +102,24 @@ def get_concordance(tokens: list, word: str, left_context_size: int, right_conte
...
-    if type(tokens) != list or type(word) != str:
-        return concordance
+    if isinstance(left_context_size, bool) and isinstance(right_context_size, bool):
+        return []
 
-    else:
-        if left_context_size >= 0 and right_context_size >= 0:
-            for i in list_index:
-                concordance.append(tokens[i - left_context_size: i + right_context_size + 1])
+    elif isinstance(tokens, list) and isinstance(word, str) and isinstance(left_context_size, int) and isinstance(
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: can you join this check and the check on line 106?

File: lab_1/main.py [hunk 18]
```diff
@@ -131,20 +137,31 @@ def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> li
...
     concordance = []
-    for i in get_concordance(tokens, word, left_n, right_n):
-        print(i)
-        concordance.append([i[0], i[-1]])
 
-    return concordance
+    if isinstance(left_n, bool) and isinstance(right_n, bool):
+        return []
+
+    elif isinstance(tokens, list) and isinstance(word, str):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: same about joining

File: lab_1/main.py [hunk 19]
```diff
@@ -131,20 +137,31 @@ def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> li
...
+        for i in get_concordance(tokens, word, left_n, right_n):
+            if isinstance(left_n, int) and isinstance(right_n, int) and left_n > 0 and right_n > 0:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: move the check of type to the beginning of this function

File: lab_1/sort_concordance_test.py [hunk 20]
```diff
@@ -108,7 +109,7 @@ def test_big_text_get_and_sort_concordance_term(self):
         """
         Checks if a context sorts right for a given term and can be found properly
         """
-        text = read_from_file('lab_1/data.txt')
+        text = read_from_file('data.txt')
```
Comment (ValeriyaKuznetsova) [inline lab_1/sort_concordance_test.py]: change back (everywhere)

File: lab_1/start.py [hunk 21]
```diff
@@ -2,17 +2,44 @@
...
+
+    RESULT = sorted_concordance
     # DO NOT REMOVE NEXT LINE - KEEP IT INTENTIONALLY LAST
-    assert RESULT == [(), ()], 'Concordance not working'
+    assert RESULT == [['corresponds', 'assigning', 'book', 'a', 'unique', 'identifier'],
+                      ['book', 'refer', 'book', 'a', 'relative', 'victim'],
+                      ['aerospace', 'source', 'book', 'a', 'village',
+                       'voice']], 'Concordance not working'  # result == то, что ожидаем от функции?
```
Comment (ValeriyaKuznetsova) [inline lab_1/start.py]: correct

File: lab_1/tokenize_test.py [hunk 22]
```diff
@@ -59,7 +60,7 @@ def test_tokenize_big_text_case(self):
         """
         Tokenize big input text scenario
         """
-        text = read_from_file('lab_1/tokens.txt')
+        text = read_from_file('tokens.txt')
```
Comment (ValeriyaKuznetsova) [inline lab_1/tokenize_test.py]: back

File: lab_1/main.py [hunk 23]
```diff
@@ -137,22 +133,20 @@ def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> li
...
     """
     concordance = []
+    statements = [isinstance(tokens, list), isinstance(word, str), not isinstance(left_n, bool),
+                  not isinstance(right_n, bool)]
 
-    if isinstance(left_n, bool) and isinstance(right_n, bool):
-        return []
-
-    elif isinstance(tokens, list) and isinstance(word, str):
+    if all(statements) and (isinstance(left_n, int) or isinstance(right_n, int)):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: can you add checks about int to statements?

**Comment by ValeriyaKuznetsova** [comment]
2 lines to change
everything else seems good!

**Comment by ValeriyaKuznetsova** [comment]
Open another PR with the second lab

**Pull request closed without merging**